File: examples/read_latency_experiment.cpp
#include "onfi_interface.hpp"
#include "onfi/timed_commands.hpp"

#include <cmath>
#include <cstdint>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <limits>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

struct Range {
    unsigned start = 0;
    unsigned count = 0;
    bool specified = false;
};

struct ExperimentConfig {
    Range blocks{};
    Range pages{};
    std::string output_path;
};

struct MeasurementRow {
    unsigned block = 0;
    unsigned page = 0;
    double t_read_us = std::numeric_limits<double>::quiet_NaN();
    uint8_t status = 0;
    std::string error;
};

bool parse_unsigned(const std::string& token, unsigned& out) {
    if (token.empty()) {
        return false;
    }
    std::size_t idx = 0;
    try {
        const unsigned long value = std::stoul(token, &idx, 0);
        if (idx != token.size() || value > std::numeric_limits<unsigned>::max()) {
            return false;
        }
        out = static_cast<unsigned>(value);
        return true;
    } catch (const std::exception&) {
        return false;
    }
}

bool parse_range(const std::string& arg, Range& range) {
    const std::size_t sep = arg.find_first_of(":,");
    const std::string first = arg.substr(0, sep);
    unsigned start = 0;
    if (!parse_unsigned(first, start)) {
        return false;
    }
    unsigned count = 1;
    if (sep != std::string::npos) {
        const std::string second = arg.substr(sep + 1);
        if (!parse_unsigned(second, count)) {
            return false;
        }
    }
    if (count == 0) {
        return false;
    }
    range.start = start;
    range.count = count;
    range.specified = true;
    return true;
}

void print_usage(const char* argv0) {
    std::cout << "Usage: sudo " << argv0
              << " [--blocks start[:count]] [--pages start[:count]]\n"
                 "                [--output file.csv]\n\n"
                 "Times ONFI page reads (tR) across the selected region; defaults to the full device.\n";
}

std::string describe_error(const onfi::timed::OperationTiming& timing) {
    if (!timing.busy_detected) {
        return "rb_not_asserted";
    }
    if (timing.timed_out) {
        return "rb_timeout";
    }
    if (!timing.ready) {
        return "status_not_ready";
    }
    if (!timing.pass) {
        return "status_fail";
    }
    return {};
}

bool parse_args(int argc, char** argv, ExperimentConfig& config, bool& show_help) {
    for (int i = 1; i < argc; ++i) {
        const std::string arg = argv[i];
        if (arg == "--blocks") {
            if (i + 1 >= argc || !parse_range(argv[++i], config.blocks)) {
                std::cerr << "Invalid --blocks argument\n";
                return false;
            }
        } else if (arg == "--pages") {
            if (i + 1 >= argc || !parse_range(argv[++i], config.pages)) {
                std::cerr << "Invalid --pages argument\n";
                return false;
            }
        } else if (arg == "--output") {
            if (i + 1 >= argc) {
                std::cerr << "--output requires a path\n";
                return false;
            }
            config.output_path = argv[++i];
        } else if (arg == "--help" || arg == "-h") {
            show_help = true;
            return false;
        } else {
            std::cerr << "Unknown argument: " << arg << "\n";
            return false;
        }
    }
    return true;
}

void ensure_range_fits(const Range& range, unsigned max_value, const char* label) {
    if (range.start >= max_value) {
        std::ostringstream oss;
        oss << label << " start " << range.start << " exceeds geometry (" << max_value << ")";
        throw std::out_of_range(oss.str());
    }
    if (range.start + range.count > max_value) {
        std::ostringstream oss;
        oss << label << " span (" << range.start << "+" << range.count << ") exceeds geometry (" << max_value << ")";
        throw std::out_of_range(oss.str());
    }
}

void write_results(std::ostream& out, const std::vector<MeasurementRow>& rows) {
    out << "block,page,t_read_us,status,error\n";
    out << std::fixed << std::setprecision(3);
    for (const auto& row : rows) {
        out << row.block << ',' << row.page << ',';
        if (!row.error.empty() || !std::isfinite(row.t_read_us)) {
            out << ',';
        } else {
            out << row.t_read_us << ',';
        }
        out << "0x" << std::uppercase << std::hex << std::setw(2) << std::setfill('0')
            << static_cast<int>(row.status) << std::nouppercase << std::dec << std::setfill(' ') << ','
            << row.error << '\n';
    }
}

} // namespace

int main(int argc, char** argv) {
    ExperimentConfig config;
    bool show_help = false;
    if (!parse_args(argc, argv, config, show_help)) {
        print_usage(argv[0]);
        return show_help ? 0 : 1;
    }

    try {
        onfi_interface onfi;
        onfi.get_started();

        if (!config.blocks.specified) {
            config.blocks.start = 0;
            config.blocks.count = onfi.num_blocks;
        }
        if (!config.pages.specified) {
            config.pages.start = 0;
            config.pages.count = onfi.num_pages_in_block;
        }

        ensure_range_fits(config.blocks, onfi.num_blocks, "Block");
        ensure_range_fits(config.pages, onfi.num_pages_in_block, "Page");

        std::ostream* stream = &std::cout;
        std::ofstream output_file;
        if (!config.output_path.empty()) {
            output_file.open(config.output_path, std::ios::out | std::ios::trunc);
            if (!output_file) {
                std::cerr << "Failed to open " << config.output_path << " for writing\n";
                return 1;
            }
            stream = &output_file;
        }

        std::vector<MeasurementRow> rows;
        rows.reserve(static_cast<std::size_t>(config.blocks.count) * config.pages.count);
        std::size_t success = 0;

        for (unsigned b = 0; b < config.blocks.count; ++b) {
            const unsigned block = config.blocks.start + b;
            for (unsigned p = 0; p < config.pages.count; ++p) {
                const unsigned page = config.pages.start + p;

                MeasurementRow row{};
                row.block = block;
                row.page = page;
                try {
                    const auto timing = onfi::timed::read_page(onfi,
                                                               block,
                                                               page,
                                                               nullptr,
                                                               0,
                                                               false,
                                                               false,
                                                               false);
                    row.status = timing.status;
                    row.error = describe_error(timing);
                    if (row.error.empty()) {
                        row.t_read_us = static_cast<double>(timing.duration_ns) / 1000.0;
                        ++success;
                    }
                } catch (const std::exception& ex) {
                    row.error = ex.what();
                }
                rows.push_back(row);
            }
        }

        write_results(*stream, rows);
        std::cerr << "Measured " << rows.size() << " pages (" << success << " succeeded)\n";
    } catch (const std::exception& ex) {
        std::cerr << "Experiment failed: " << ex.what() << "\n";
        return 1;
    }

    return 0;
}

File: examples/scripts/read_latency_experiment.lua
--[=[
Measure ONFI page read latency (tR) across a region using nandworks.timed bindings.

Usage:
  sudo bin/nandworks script --allow-unsafe examples/scripts/read_latency_experiment.lua \
       [--blocks start[:count]] [--pages start[:count]] [--output file.csv] [--include-spare]

Defaults to the entire device if no ranges are provided. CSV is written to stdout unless
--output is set (requires --allow-unsafe to enable file I/O).
]=]

local function usage()
  print("Usage: sudo bin/nandworks script --allow-unsafe examples/scripts/read_latency_experiment.lua \\")
  print("         [--blocks start[:count]] [--pages start[:count]] [--output file.csv] [--include-spare]")
end

local function parse_range(text)
  local sep = string.find(text, "[:,]")
  local first = sep and string.sub(text, 1, sep - 1) or text
  local second = sep and string.sub(text, sep + 1) or nil

  local start = tonumber(first)
  local count = second and tonumber(second) or 1
  if not start or start < 0 then
    return nil, "invalid start in range: " .. text
  end
  if not count or count <= 0 then
    return nil, "invalid count in range: " .. text
  end
  return { start = start, count = count }
end

local function describe_error(timing)
  if not timing.busy_detected then
    return "rb_not_asserted"
  end
  if timing.timed_out then
    return "rb_timeout"
  end
  if not timing.ready then
    return "status_not_ready"
  end
  if not timing.pass then
    return "status_fail"
  end
  return ""
end

local function make_writer(path)
  if path then
    if not io or not io.open then
      error("File output requires --allow-unsafe to enable io.open")
    end
    local fh, err = io.open(path, "w")
    if not fh then
      error("Failed to open output '" .. tostring(path) .. "': " .. tostring(err))
    end
    return function(line) fh:write(line, "\n") end, function() fh:close() end
  end
  return function(line) print(line) end, function() end
end

local function parse_args()
  local cfg = {
    include_spare = false,
  }

  local i = 1
  while i <= #arg do
    local token = arg[i]
    if token == "--blocks" then
      i = i + 1
      if i > #arg then error("--blocks expects a value") end
      local range, err = parse_range(arg[i])
      if not range then error(err) end
      cfg.blocks = range
    elseif token == "--pages" then
      i = i + 1
      if i > #arg then error("--pages expects a value") end
      local range, err = parse_range(arg[i])
      if not range then error(err) end
      cfg.pages = range
    elseif token == "--output" then
      i = i + 1
      if i > #arg then error("--output expects a path") end
      cfg.output = arg[i]
    elseif token == "--include-spare" or token == "--include_spare" then
      cfg.include_spare = true
    elseif token == "--help" or token == "-h" then
      cfg.help = true
      break
    else
      error("Unknown argument: " .. tostring(token))
    end
    i = i + 1
  end

  return cfg
end

local function ensure_range(range, max_value, label)
  if range.start >= max_value then
    error(string.format("%s start %d exceeds geometry (%d)", label, range.start, max_value))
  end
  if range.start + range.count > max_value then
    error(string.format("%s span (%d+%d) exceeds geometry (%d)", label, range.start, range.count, max_value))
  end
end

local config = parse_args()
if config.help then
  usage()
  return
end
local include_spare = config.include_spare

nandworks.with_session(function(cmd, nw)
  local geom = nw.geometry()
  if not config.blocks then
    config.blocks = { start = 0, count = geom.blocks }
  end
  if not config.pages then
    config.pages = { start = 0, count = geom.pages_per_block }
  end

  ensure_range(config.blocks, geom.blocks, "Block")
  ensure_range(config.pages, geom.pages_per_block, "Page")

  local write_line, close_writer = make_writer(config.output)
  write_line("block,page,t_read_us,status,error")

  local rows = 0
  local successes = 0

  for b = 0, config.blocks.count - 1 do
    local block = config.blocks.start + b
    for p = 0, config.pages.count - 1 do
      local page = config.pages.start + p

      local status = 0
      local t_read_us = nil
      local error_text = ""

      local ok, timing = pcall(nw.timed.read_page, {
        block = block,
        page = page,
        include_spare = include_spare,
        fetch_data = false,
        verbose = false,
      })

      if ok then
        status = timing.status or 0
        error_text = describe_error(timing)
        if error_text == "" then
          t_read_us = (timing.duration_ns or 0) / 1000.0
          successes = successes + 1
        end
      else
        error_text = timing
      end

      if error_text ~= "" or not t_read_us then
        write_line(string.format("%d,%d,,0x%02X,%s", block, page, status, error_text))
      else
        write_line(string.format("%d,%d,%.3f,0x%02X,%s", block, page, t_read_us, status, error_text))
      end

      rows = rows + 1
    end
  end

  close_writer()
  print(string.format("Measured %d pages (%d succeeded)", rows, successes))
end)
